File: Hearthstone-Script-Deck-SDK/src/main/kotlin/club/xiaojiawei/DeckStrategy.java
package club.xiaojiawei;

import club.xiaojiawei.bean.Card;
import club.xiaojiawei.enums.RunModeEnum;

import java.util.HashSet;
import java.util.Objects;

/**
 * @author 肖嘉威
 * @date 2024/9/7 13:36
 */
public abstract class DeckStrategy {


    public final RunModeEnum[] runMode = getRunMode();

    /**
     * 卡组名，将会显示在界面中
     * @return 非空
     */
    abstract public String name();

    /**
     * 卡组运行的模式
     * @return 返回非null非空且不包含null,推荐每次返回的数组对象是一样的
     */
    abstract protected RunModeEnum[] getRunMode();

    /**
     * 卡组代码
     * @return 可空
     */
    abstract public String deckCode();

    /**
     * 卡组唯一标识
     * @return 非空，长度必需为36
     */
    abstract public String id();


    /**
     * 执行换牌策略
     * @param cards 需要换掉的牌直接从集合中删除
     */
    abstract public void executeChangeCard(HashSet<Card> cards);

    /**
     * 执行出牌策略
     */
    abstract public void executeOutCard();

    /**
     * 执行发现选牌
     * @param cards 发现的牌
     * @return 返回范围 [0,数组长度)
     */
    abstract public int executeDiscoverChooseCard(Card...cards);

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        DeckStrategy that = (DeckStrategy) o;
        return Objects.equals(id(), that.id());
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id());
    }
}

File: Hearthstone-Script-Deck-SDK/src/main/kotlin/club/xiaojiawei/DeckStrategy.kt
package club.xiaojiawei

import club.xiaojiawei.bean.Card
import club.xiaojiawei.enums.RunModeEnum
import java.util.*

/**
 * @author 肖嘉威
 * @date 2024/9/9 0:33
 */
abstract class DeckStrategy {

    var pluginId: String = ""

    val runModes: Array<RunModeEnum> by lazy { getRunMode() }

    /**
     * 是否需要投降，改为true时软件会在适合的时机投降
     */
    var needSurrender = false

    /**
     * 每局游戏开始时调用此方法
     */
    open fun reset(){
        needSurrender = false
    }

    /**
     * 卡组名，将会显示在界面中
     * @return 非空
     */
    abstract fun name(): String

    /**
     * 卡组运行的模式
     * @return 返回非null非空且不包含null,推荐每次返回的数组对象是一样的
     */
    protected abstract fun getRunMode(): Array<RunModeEnum>

    /**
     * 卡组代码
     * @return 非空
     */
    abstract fun deckCode(): String

    /**
     * 卡组唯一标识
     * @return 非空
     */
    abstract fun id(): String

    /**
     * 执行换牌策略
     * @param cards 需要换掉的牌直接从集合中删除
     */
    abstract fun executeChangeCard(cards: HashSet<Card>)

    /**
     * 执行出牌策略
     */
    abstract fun executeOutCard()

    /**
     * 执行发现选牌
     * @param cards 发现的牌
     * @return 返回范围 [0,数组长度)
     */
    abstract fun executeDiscoverChooseCard(vararg cards: Card): Int

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as DeckStrategy
        return id() == that.id()
    }

    override fun hashCode(): Int {
        return Objects.hashCode(id())
    }
}